### DISCUSSION TITLE
[React] Is it possible to use nested useDismiss? (Dialog together with the Select)

### DISCUSSION THREAD

**Discussion by Marsony** [floating-ui/floating-ui#2206] [React] [answered]
Hello everyone, Im trying to use "Select" component inside the "Dialog" component.

Both of this components I got from the floating-ui docs.

So, when Im focusing on the Select component and press "Esc", my Dialog component closes too together with the Select.

(I don't want to close my Dialog, I want to close only the Select)

Thank you very much.

**Comment by atomiks** [comment, marked as answer]
Hey, take a look here: https://github.com/floating-ui/floating-ui/discussions/2138#discussioncomment-4762775. Basically, you can use `FloatingTree` to specify how floating elements are nested, which allows the global keydown listeners attached to the document to ignore certain events.

**Reply by Marsony** [reply]
> Hey, take a look here: [#2138 (comment)](https://github.com/floating-ui/floating-ui/discussions/2138#discussioncomment-4762775). Basically, you can use `FloatingTree` to specify how floating elements are nested, which allows the global keydown listeners attached to the document to ignore certain events.

I wrap my Select in FloatingNode with key (+pass key in the useFloating) and did the same with Dialog

but it doesn't help, do you have an example? ty

**Reply by atomiks** [reply]
Here's an example: https://codesandbox.io/s/divine-glitter-o077s1?file=/src/Dialog.tsx

The main `Dialog`'s `useDialog`, which sets up its `nodeId`, needs to be called within the `FloatingTree`, like so:
```js
export function DialogComponent({
  children,
  ...options
}: {
  children: React.ReactNode;
} & DialogOptions) {
  const dialog = useDialog(options);
  return (
    <DialogContext.Provider value={dialog}>{children}</DialogContext.Provider>
  );
}

export function Dialog(
  props: {
    children: React.ReactNode;
  } & DialogOptions
) {
  return (
    <FloatingTree>
      <DialogComponent {...props} />
    </FloatingTree>
  );
}
```

A `nodeId` is passed to both `useFloating` hooks, as well as `FloatingNode` wrapping the floating element, with `bubbles: false` passed to the Select's `useDismiss` hook.

---

If all else fails, you can manually stop propagation like this:

```js
<div 
  {...getFloatingProps({
    onKeyDown: e => {
      if (e.key === 'Escape' && context.open) {
        e.stopPropagation();
        context.setOpen(false);
      }
    }
  })} 
/>
```

I do want to make this easier as it's definitely a pain point (automatic using the React tree if possible). Going to look into it shortly.

**Reply by Marsony** [reply]
Excellent)!
